include/internal/ktls.h

```diff
@@ -309,6 +309,103 @@ static ossl_inline int ktls_read_record(int fd, void *data, size_t length)
 }
 
 #   endif /* OPENSSL_NO_KTLS_RX */
+
+/* Function to check supported ciphers in Linux */
+static ossl_inline int ktls_check_supported_cipher(const EVP_CIPHER *c,
+                                            const EVP_CIPHER_CTX *dd)
+{
+    /* check that cipher is AES_GCM_128, AES_GCM_256, AES_CCM_128 */
+    switch (EVP_CIPHER_nid(c))
+    {
+#  ifdef OPENSSL_KTLS_AES_CCM_128
+    case NID_aes_128_ccm:
+        if (EVP_CIPHER_CTX_tag_length(dd) != EVP_CCM_TLS_TAG_LEN)
+          return 0;
+#  endif
+#  ifdef OPENSSL_KTLS_AES_GCM_128
+    case NID_aes_128_gcm:
+#  endif
+#  ifdef OPENSSL_KTLS_AES_GCM_256
+    case NID_aes_256_gcm:
+#  endif
+        return 1;
+    default:
+        return 0;
+    }
+}
+
+/* Function to configure kernel TLS structure */
+static ossl_inline int ktls_configure_crypto(const EVP_CIPHER *c, int tls_version,
+                                EVP_CIPHER_CTX *dd, void *rl_sequence,
+                                struct tls_crypto_info_all *crypto_info,
+                                unsigned char **rec_seq, unsigned char *iv,
+                                unsigned char *key)
+{
+    unsigned char geniv[12];
+    unsigned char *iiv = iv;
+
+    if (tls_version == TLS1_2_VERSION &&
+        EVP_CIPHER_mode(c) == EVP_CIPH_GCM_MODE) {
+        EVP_CIPHER_CTX_ctrl(dd, EVP_CTRL_GET_IV,
+                            EVP_GCM_TLS_FIXED_IV_LEN + EVP_GCM_TLS_EXPLICIT_IV_LEN,
+                            geniv);
+        iiv = geniv;
+    }
```

Comment:
@mattcaswell I mean this part of code, it was taken from initial commit of Kernel TLS

Reply:
It looks like it is reading the value OpenSSL computed for the explicit part of the IV in TLS 1.2.  The IV for GCM (and I believe for CCM) is split into an implicit salt computed from the PRF, an explicit 8 byte counter that increments for each TLS PDU, and a 4-byte block counter that increments for each block within a PDU.  The 8 byte explicit value is on the wire, so the sender is free to choose any value with the constraint as I understand that the explicit IV must be unique and never reused with the same session key and salt.  One option permitted by the RFC is to reuse the TLS sequence number as the explicit IV.  OpenSSL generates a random 8 byte value that it then increments for each TLS PDU.  In FreeBSD's KTLS, we only pass in the salt from userland to the kernel and the kernel then generates its own random starting value that it increments for each TLS PDU.  It seems here that the existing code is reusing the random value generated by OpenSSL for the starting value of the explicit IV rather than relying on the kernel to generate its own random starting value (or just using the TLS record sequence number as the explicit IV).

So, the real question is how does the Linux kernel expect to deal with the explicit 8 byte IV for TLS 1.2?  Does it expect the socket option to initialize the starting value, or does it generate its own?  As it is, right now for CCM you are passing the next 8 bytes of the PRF result after the IV as the explicit IV's starting value.  That might be ok as long as the kernel keeps incrementing the value, (and the value is public on the wire anyway), but technically you are reading beyond the bounds of 'iiv'.  Most likely you need to fetch the full IV for CCM as well for TLS 1.2 to avoid that out-of-bounds read.

For TLS 1.3, the entire 12 byte nonce is implicit from the PRF.

Reply:
John, thank you very much for clarification. I have found the place in the code where explicit part of IV is filled with random data in case of GCM (it is in EVP_CTRL_GCM_SET_IV_FIXED section), and that's it should be read. But for CCM mode only implicit salt is copied to CIPHER_CTX and explicit part is untouched. The iv is taken from s->s3.tmp.key_block, and that key_block is allocated and computed from PRF. The size of s->s3.tmp.key_block is computed with full size of IV, so there is no out-of-bounds read, and the value is permitted by the RFC 6655.
The Linux kernel expects that socket option must initialize the starting value and then it just keeps incrementing it.
So, again, thank you very much for clarification, now I fully understand the need for this part of code.